File: flowersec-go/controlplane/channelinit/channelinit.go
package channelinit

import (
	"crypto/rand"
	"encoding/json"
	"errors"
	"math"
	"time"

	"github.com/floegence/flowersec/flowersec-go/controlplane/issuer"
	"github.com/floegence/flowersec/flowersec-go/controlplane/token"
	controlv1 "github.com/floegence/flowersec/flowersec-go/gen/flowersec/controlplane/v1"
	e2eev1 "github.com/floegence/flowersec/flowersec-go/gen/flowersec/e2ee/v1"
	"github.com/floegence/flowersec/flowersec-go/internal/base64url"
	"github.com/floegence/flowersec/flowersec-go/internal/timeutil"
)

const (
	// ChannelInitWindowSeconds bounds how long a grant remains valid.
	ChannelInitWindowSeconds = 120
	// DefaultIdleTimeoutSeconds advertises the tunnel idle timeout to endpoints.
	//
	// This value is embedded into signed tokens and enforced by the tunnel.
	DefaultIdleTimeoutSeconds = 60
	// DefaultTokenExpSeconds is used when TokenExpSeconds is unset.
	DefaultTokenExpSeconds = 60
)

var ErrChannelInitExpired = errors.New("channel init expired")

// Params define channel-init issuance settings and defaults.
type Params struct {
	TunnelURL      string // WebSocket URL for tunnel server.
	TunnelAudience string // Expected audience for issued tokens.
	IssuerID       string // Issuer identifier embedded in tokens.

	TokenExpSeconds    int64         // Token lifetime in seconds (0 uses default; capped by init exp).
	IdleTimeoutSeconds int32         // Tunnel idle timeout enforced per channel (seconds) (0 uses default).
	ClockSkew          time.Duration // Allowed clock skew for validation hints.

	AllowedSuites []e2eev1.Suite // E2EE suites permitted for the channel.
	DefaultSuite  e2eev1.Suite   // Default E2EE suite for the channel.
}

// Service issues channel-init grants and tokens for clients/servers.
type Service struct {
	Issuer *issuer.Keyset   // Signing keyset for tunnel tokens.
	Params Params           // Defaults and limits for channel-init grants.
	Now    func() time.Time // Optional time source override.
}

func (s *Service) now() time.Time {
	if s.Now != nil {
		return s.Now()
	}
	return time.Now()
}

// NewChannelInit creates paired client/server grants with shared PSK and tokens.
func (s *Service) NewChannelInit(channelID string) (client *controlv1.ChannelInitGrant, server *controlv1.ChannelInitGrant, err error) {
	if s.Issuer == nil {
		return nil, nil, errors.New("missing issuer")
	}
	if s.Params.TunnelURL == "" {
		return nil, nil, errors.New("missing tunnel url")
	}
	if s.Params.TunnelAudience == "" {
		return nil, nil, errors.New("missing tunnel audience")
	}
	if s.Params.IssuerID == "" {
		return nil, nil, errors.New("missing issuer id")
	}
	if channelID == "" {
		return nil, nil, errors.New("missing channel_id")
	}
	psk, err := randomBytes(32)
	if err != nil {
		return nil, nil, err
	}
	pskB64u := base64url.Encode(psk)

	now := s.now()
	initExp := now.Add(ChannelInitWindowSeconds * time.Second).Unix()
	tokenExpSeconds := s.Params.TokenExpSeconds
	if tokenExpSeconds < 0 {
		return nil, nil, errors.New("token_exp_seconds must be >= 0")
	}
	if tokenExpSeconds == 0 {
		tokenExpSeconds = DefaultTokenExpSeconds
	}
	idleTimeoutSeconds := s.Params.IdleTimeoutSeconds
	if idleTimeoutSeconds < 0 {
		return nil, nil, errors.New("idle_timeout_seconds must be >= 0")
	}
	if idleTimeoutSeconds == 0 {
		idleTimeoutSeconds = DefaultIdleTimeoutSeconds
	}

	allowedSuitesE2EE := s.Params.AllowedSuites
	if len(allowedSuitesE2EE) == 0 {
		allowedSuitesE2EE = []e2eev1.Suite{e2eev1.Suite_X25519_HKDF_SHA256_AES_256_GCM}
	}
	allowedSuitesE2EE = normalizeSuites(allowedSuitesE2EE)
	if len(allowedSuitesE2EE) == 0 {
		return nil, nil, errors.New("no allowed suites")
	}
	allowedSuites := make([]controlv1.Suite, 0, len(allowedSuitesE2EE))
	for _, s := range allowedSuitesE2EE {
		allowedSuites = append(allowedSuites, controlv1.Suite(s))
	}
	defaultSuiteE2EE := s.Params.DefaultSuite
	if defaultSuiteE2EE == 0 {
		// If the caller does not specify a default, prefer the first allowed suite.
		defaultSuiteE2EE = allowedSuitesE2EE[0]
	}
	if !containsSuite(allowedSuitesE2EE, defaultSuiteE2EE) {
		return nil, nil, errors.New("default suite not allowed")
	}
	defaultSuite := controlv1.Suite(defaultSuiteE2EE)

	clientToken, err := s.signRoleToken(channelID, uint8(controlv1.Role_client), initExp, idleTimeoutSeconds, tokenExpSeconds, now)
	if err != nil {
		return nil, nil, err
	}
	serverToken, err := s.signRoleToken(channelID, uint8(controlv1.Role_server), initExp, idleTimeoutSeconds, tokenExpSeconds, now)
	if err != nil {
		return nil, nil, err
	}

	client = &controlv1.ChannelInitGrant{
		TunnelUrl:                s.Params.TunnelURL,
		ChannelId:                channelID,
		ChannelInitExpireAtUnixS: initExp,
		IdleTimeoutSeconds:       idleTimeoutSeconds,
		Role:                     controlv1.Role_client,
		Token:                    clientToken,
		E2eePskB64u:              pskB64u,
		AllowedSuites:            allowedSuites,
		DefaultSuite:             defaultSuite,
	}
	server = &controlv1.ChannelInitGrant{
		TunnelUrl:                s.Params.TunnelURL,
		ChannelId:                channelID,
		ChannelInitExpireAtUnixS: initExp,
		IdleTimeoutSeconds:       idleTimeoutSeconds,
		Role:                     controlv1.Role_server,
		Token:                    serverToken,
		E2eePskB64u:              pskB64u,
		AllowedSuites:            allowedSuites,
		DefaultSuite:             defaultSuite,
	}
	return client, server, nil
}

// ReissueToken refreshes the signed token while keeping the same grant fields.
func (s *Service) ReissueToken(grant *controlv1.ChannelInitGrant) (*controlv1.ChannelInitGrant, error) {
	if s.Issuer == nil {
		return nil, errors.New("missing issuer")
	}
	if s.Params.TunnelAudience == "" {
		return nil, errors.New("missing tunnel audience")
	}
	if s.Params.IssuerID == "" {
		return nil, errors.New("missing issuer id")
	}
	if grant == nil {
		return nil, errors.New("missing grant")
	}
	if grant.IdleTimeoutSeconds <= 0 {
		return nil, errors.New("missing idle_timeout_seconds")
	}
	now := s.now()
	skew := s.Params.ClockSkew
	if skew < 0 {
		skew = 0
	}
	skew = timeutil.NormalizeSkew(skew)
	if now.Unix() > timeutil.AddSkewUnix(grant.ChannelInitExpireAtUnixS, skew) {
		return nil, ErrChannelInitExpired
	}
	tokenExpSeconds := s.Params.TokenExpSeconds
	if tokenExpSeconds < 0 {
		return nil, errors.New("token_exp_seconds must be >= 0")
	}
	if tokenExpSeconds == 0 {
		tokenExpSeconds = DefaultTokenExpSeconds
	}
	role := uint8(grant.Role)
	newToken, err := s.signRoleToken(grant.ChannelId, role, grant.ChannelInitExpireAtUnixS, grant.IdleTimeoutSeconds, tokenExpSeconds, now)
	if err != nil {
		return nil, err
	}
	out := *grant
	out.Token = newToken
	return &out, nil
}

func (s *Service) signRoleToken(channelID string, role uint8, initExp int64, idleTimeoutSeconds int32, tokenExpSeconds int64, now time.Time) (string, error) {
	tokenID, err := randomB64u(24)
	if err != nil {
		return "", err
	}
	iat := now.Unix()
	exp := iat
	if tokenExpSeconds > 0 {
		// Avoid time.Duration overflow when tokenExpSeconds is very large.
		if iat > math.MaxInt64-tokenExpSeconds {
			exp = math.MaxInt64
		} else {
			exp = iat + tokenExpSeconds
		}
	}
	if exp > initExp {
		exp = initExp
	}
	return s.Issuer.SignToken(token.Payload{
		Aud:                s.Params.TunnelAudience,
		Iss:                s.Params.IssuerID,
		ChannelID:          channelID,
		Role:               role,
		TokenID:            tokenID,
		InitExp:            initExp,
		IdleTimeoutSeconds: idleTimeoutSeconds,
		Iat:                iat,
		Exp:                exp,
	})
}

func randomB64u(n int) (string, error) {
	b, err := randomBytes(n)
	if err != nil {
		return "", err
	}
	return base64url.Encode(b), nil
}

func randomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	_, err := rand.Read(b)
	return b, err
}

// MarshalGrantJSON encodes the grant for transport to the client.
func MarshalGrantJSON(g *controlv1.ChannelInitGrant) ([]byte, error) {
	return json.Marshal(g)
}

func normalizeSuites(in []e2eev1.Suite) []e2eev1.Suite {
	if len(in) == 0 {
		return nil
	}
	out := make([]e2eev1.Suite, 0, len(in))
	seen := make(map[e2eev1.Suite]struct{}, len(in))
	for _, s := range in {
		if s == 0 {
			continue
		}
		if _, ok := seen[s]; ok {
			continue
		}
		seen[s] = struct{}{}
		out = append(out, s)
	}
	return out
}

func containsSuite(list []e2eev1.Suite, want e2eev1.Suite) bool {
	for _, s := range list {
		if s == want {
			return true
		}
	}
	return false
}

File: flowersec-ts/src/controlplane/channelInit.ts
import type { ChannelInitGrant, Suite } from "../gen/flowersec/controlplane/v1.gen.js";
import { Role } from "../gen/flowersec/controlplane/v1.gen.js";
import { base64urlEncode } from "../utils/base64url.js";
import type { IssuerKeyset } from "./issuer.js";

export const CHANNEL_INIT_WINDOW_SECONDS = 120;
export const DEFAULT_IDLE_TIMEOUT_SECONDS = 60;
export const DEFAULT_TOKEN_EXP_SECONDS = 60;

export type ChannelInitParams = Readonly<{
  tunnelUrl: string;
  tunnelAudience: string;
  issuerId: string;
  tokenExpSeconds?: number;
  idleTimeoutSeconds?: number;
  clockSkewMs?: number;
  allowedSuites?: readonly Suite[];
  defaultSuite?: Suite;
}>;

export class ChannelInitService {
  constructor(
    private readonly issuer: IssuerKeyset,
    private readonly params: ChannelInitParams,
    private readonly nowUnixS: () => number = () => Math.floor(Date.now() / 1000),
  ) {}

  issue(channelIdInput: string): Readonly<{ client: ChannelInitGrant; server: ChannelInitGrant }> {
    const params = normalizeParams(this.params);
    const channelId = normalizeChannelId(channelIdInput);
    const psk = crypto.getRandomValues(new Uint8Array(32));
    let pskB64u: string;
    try {
      pskB64u = base64urlEncode(psk);
    } finally {
      psk.fill(0);
    }
    const now = normalizeNow(this.nowUnixS());
    const initExp = now + CHANNEL_INIT_WINDOW_SECONDS;
    const grant = (role: Role, token: string): ChannelInitGrant => ({
      tunnel_url: params.tunnelUrl,
      channel_id: channelId,
      channel_init_expire_at_unix_s: initExp,
      idle_timeout_seconds: params.idleTimeoutSeconds,
      role,
      token,
      e2ee_psk_b64u: pskB64u,
      allowed_suites: [...params.allowedSuites],
      default_suite: params.defaultSuite,
    });
    return {
      client: grant(Role.Role_client, this.signRoleToken(channelId, Role.Role_client, initExp, params.idleTimeoutSeconds, params.tokenExpSeconds, now, params)),
      server: grant(Role.Role_server, this.signRoleToken(channelId, Role.Role_server, initExp, params.idleTimeoutSeconds, params.tokenExpSeconds, now, params)),
    };
  }

  reissue(grant: ChannelInitGrant): ChannelInitGrant {
    const params = normalizeParams(this.params);
    if (grant.idle_timeout_seconds <= 0 || (grant.role !== Role.Role_client && grant.role !== Role.Role_server)) throw new Error("invalid grant");
    const now = normalizeNow(this.nowUnixS());
    const skewSeconds = Math.ceil(params.clockSkewMs / 1000);
    if (now > grant.channel_init_expire_at_unix_s + skewSeconds) throw new Error("channel init expired");
    return {
      ...grant,
      token: this.signRoleToken(
        normalizeChannelId(grant.channel_id),
        grant.role,
        grant.channel_init_expire_at_unix_s,
        grant.idle_timeout_seconds,
        params.tokenExpSeconds,
        now,
        params,
      ),
    };
  }

  private signRoleToken(
    channelId: string,
    role: Role,
    initExp: number,
    idleTimeoutSeconds: number,
    tokenExpSeconds: number,
    now: number,
    params: ReturnType<typeof normalizeParams>,
  ): string {
    const iat = Math.min(now, initExp);
    const exp = Math.min(initExp, iat + tokenExpSeconds);
    const tokenId = crypto.getRandomValues(new Uint8Array(24));
    try {
      return this.issuer.sign({
        aud: params.tunnelAudience,
        iss: params.issuerId,
        channel_id: channelId,
        role: role as 1 | 2,
        token_id: base64urlEncode(tokenId),
        init_exp: initExp,
        idle_timeout_seconds: idleTimeoutSeconds,
        iat,
        exp,
      });
    } finally {
      tokenId.fill(0);
    }
  }
}

function normalizeParams(params: ChannelInitParams) {
  const tunnelUrl = params.tunnelUrl.trim();
  const tunnelAudience = params.tunnelAudience.trim();
  const issuerId = params.issuerId.trim();
  if (tunnelUrl === "") throw new Error("missing tunnel URL");
  if (tunnelAudience === "") throw new Error("missing tunnel audience");
  if (issuerId === "") throw new Error("missing issuer ID");
  const tokenExpSeconds = normalizeNonNegative("tokenExpSeconds", params.tokenExpSeconds ?? 0) || DEFAULT_TOKEN_EXP_SECONDS;
  const idleTimeoutSeconds = normalizeNonNegative("idleTimeoutSeconds", params.idleTimeoutSeconds ?? 0) || DEFAULT_IDLE_TIMEOUT_SECONDS;
  const clockSkewMs = normalizeNonNegative("clockSkewMs", params.clockSkewMs ?? 0);
  const allowedSuites = [...new Set(params.allowedSuites?.length ? params.allowedSuites : [1 as Suite])];
  for (const suite of allowedSuites) if (suite !== 1 && suite !== 2) throw new Error("unsupported suite");
  const defaultSuite = params.defaultSuite ?? allowedSuites[0]!;
  if (!allowedSuites.includes(defaultSuite)) throw new Error("default suite not allowed");
  return { tunnelUrl, tunnelAudience, issuerId, tokenExpSeconds, idleTimeoutSeconds, clockSkewMs, allowedSuites, defaultSuite };
}

function normalizeChannelId(input: string): string {
  const value = input.trim();
  if (value === "" || new TextEncoder().encode(value).length > 256) throw new Error("invalid channel ID");
  return value;
}

function normalizeNonNegative(name: string, value: number): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new Error(`${name} must be a non-negative integer`);
  return value;
}

function normalizeNow(value: number): number {
  if (!Number.isSafeInteger(value) || value <= 0) throw new Error("invalid clock");
  return value;
}
